src/main/java/io/kamax/matrix/client/regular/MatrixHttpClient.java

```diff
@@ -63,4 +63,34 @@ public _MatrixUser getUser(_MatrixID mxId) {
         return new MatrixHttpUser(getContext(), mxId);
     }
 
+    @Override
+    public Optional<String> getDeviceId() {
+        return context.getDeviceId();
+    }
+
+    @Override
+    public void login(MatrixPasswordLoginCredentials credentials) {
+        HttpPost request = new HttpPost(getClientPath("/login"));
+        if (context.getDeviceId().isPresent()) {
+            request.setEntity(getJsonEntity(new LoginPostBody(credentials.getLocalPart(), credentials.getPassword(),
+                    context.getDeviceId().get())));
+        } else {
+            request.setEntity(getJsonEntity(new LoginPostBody(credentials.getLocalPart(), credentials.getPassword())));
+        }
+
+        String body = execute(request);
+        LoginResponse response = gson.fromJson(body, LoginResponse.class);
+        context.setToken(response.getAccessToken());
+        context.setDeviceId(response.getDeviceId());
+        context.setUser(new MatrixID(credentials.getLocalPart(), context.getHs().getDomain()));
```

Comment:
[As per spec](https://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-r0-login), the Matrix ID of the user is returned in the response body, so we should not build it ourself.